### PR TITLE
refactor debug

### PR DESCRIPTION
**Pull request by alvarofe**
Don't merge yet. I will try to review more code and refactor as much as I can. 

Don't hesitate to comment if you see something wrong :)

### CODE REVIEW

File: libr/debug/p/debug_native.c [hunk 1]
```diff
@@ -1083,51 +1087,36 @@ static int r_debug_native_reg_read(RDebug *dbg, int type, ut8 *buf, int size) {
...
+	thread_get_state (inferior_threads[tid], (state), (thread_state_t)regs, &gp_count)
+
+static int xnu_reg_read (RDebug *dbg, int type, ut8 *buf, int size) {
 	int ret;
+	int pid = dbg->pid;
 	thread_array_t inferior_threads = NULL;
 	unsigned int inferior_thread_count = 0;
 	R_DEBUG_REG_T *regs = (R_DEBUG_REG_T*)buf;
-        unsigned int gp_count = R_DEBUG_STATE_SZ; //sizeof (R_DEBUG_REG_T);
+    unsigned int gp_count = R_DEBUG_STATE_SZ;
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: wrong indentation

File: libr/debug/p/debug_native.c [hunk 2]
```diff
@@ -1083,51 +1087,36 @@ static int r_debug_native_reg_read(RDebug *dbg, int type, ut8 *buf, int size) {
...
 	R_DEBUG_REG_T *regs = (R_DEBUG_REG_T*)buf;
-        unsigned int gp_count = R_DEBUG_STATE_SZ; //sizeof (R_DEBUG_REG_T);
+    unsigned int gp_count = R_DEBUG_STATE_SZ; 
 	int tid = dbg->tid;
 
-#if 0
-	// if uncommented, it will break x86-32 debugging from x86-64 (ios simulator f.ex)
-	if (size<sizeof (R_DEBUG_REG_T)) {
-		eprintf ("Small buffer passed to r_debug_read\n");
+    ret = task_threads (pid_to_task (pid), &inferior_threads, &inferior_thread_count);
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: wrong indentation here and on the next line

File: libr/debug/p/debug_native.c [hunk 3]
```diff
@@ -1135,203 +1124,128 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
-#endif
-#endif
-		return R_TRUE;
-		break;
-	case R_REG_TYPE_FPU:
-	case R_REG_TYPE_MMX:
-	case R_REG_TYPE_XMM:
-#if __linux__
+	} else eprintf ("There are no threads!\n");
+        return sizeof (R_DEBUG_REG_T);
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: wrong indentation

File: libr/debug/p/debug_native.c [hunk 4]
```diff
@@ -1135,203 +1124,128 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
-			eprintf ("mxcr_mask = 0x%08x\n", fpregs.mxcr_mask);
-			for(i=0;i<8;i++) {
+	#if !__ANDROID__
+		eprintf ("---- x86-64 ----\n");
+		PRINT_FPU(fpregs);
+		eprintf ("size = 0x%08x\n", (ut32)sizeof (fpregs));
+		for (i=0;i<16;i++) {
+			ut32 *a = (ut32*)&fpregs.xmm_space;
+			a = a + (i * 4);
+			eprintf ("xmm%d = %08x %08x %08x %08x   ",i
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: be consistent with the spaces after commas

File: libr/debug/p/debug_native.c [hunk 5]
```diff
@@ -1135,203 +1124,128 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
 				eprintf ("st%d =%0.3lg (0x%016"PFMT64x") | %0.3f (%08x)  | %0.3f (%08x) \n", i
-						,(double)*((double*)&fpregs.st_space[i*4])
-						,*b
-						,(float) f[0]
-						,c[0]
-						,(float) f[1]
-						,c[1]
-					);
-			}
+					,(double)*((double*)&fpregs.st_space[i*4]), *b, (float)f[0], c[0], (float)f[1], c[1]);
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: space after comma

File: libr/debug/p/debug_native.c [hunk 6]
```diff
@@ -1135,203 +1124,128 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
-		if (ret1 != 0)
-			return R_FALSE;
-		if (sizeof (fpregs) < size)
-			size = sizeof (fpregs);
-		memcpy (buf, &fpregs, size);
-		return sizeof (fpregs)
-#endif
+	#else
+		PRINT_FPU(fpregs);
+		for(i=0;i<8;i++) {
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: space after for

File: libr/debug/p/debug_native.c [hunk 7]
```diff
@@ -1135,203 +1124,128 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
+			ut64 *b = (ut64 *)&fpregs.st_space[i*4];
+			ut32 *c =(ut32*)&fpregs.st_space;
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: consistency in spaces before/after '='

Comment (ret2libc) [inline libr/debug/p/debug_native.c]: and also on the next line

File: libr/debug/p/debug_native.c [hunk 8]
```diff
@@ -1135,203 +1124,128 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
+			float *f=(float *)&fpregs.st_space;
+			c=c+(i*4);
+			f=f+(i*4);
+			eprintf ("st%d =%0.3lg (0x%016"PFMT64x") | %0.3f (%08x)  | %0.3f (%08x) \n", i
+				,(double)*((double*)&fpregs.st_space[i*4]),*b,(float) f[0], c[0], (float) f[1], c[1]);
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: spaces after commas

File: libr/debug/p/debug_native.c [hunk 9]
```diff
@@ -1344,100 +1258,144 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
 					eprintf ("st%d =%0.3lg (0x%016"PFMT64x") | %0.3f (0x%08x)  | %0.3f (0x%08x)\n", i
-							,(double)*((double*)(&fpxregs.st_space[i*4]))
-							,b[0]
-							,f[0]
-							,c[0]
-							,f[1]
-							,c[1]
-							);
-				}
+						,(double)*((double*)(&fpxregs.st_space[i*4])),b[0], f[0], c[0], f[1], c[1]);
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: spaces after commas

File: libr/debug/p/debug_native.c [hunk 10]
```diff
@@ -1344,100 +1258,144 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
-			memcpy (buf, &fpregs, size);
-			return sizeof (fpregs);
-#endif
-#endif
+			#endif // !__ANDROID__
+				return;
 		}
+		eprintf ("---- x86-32-noxmm ----\n");
+		PRINT_FPU_NOXMM(fpregs);
+		for(i=0;i<8;i++) {
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: space after for

File: libr/debug/p/debug_native.c [hunk 11]
```diff
@@ -1344,100 +1258,144 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
+			ut64 *b = (ut64 *)(&fpregs.st_space[i*4]);
+			double *d = (double*)b;
+			ut32 *c =(ut32*)&fpregs.st_space;
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: spaces around '='

File: libr/debug/p/debug_native.c [hunk 12]
```diff
@@ -1344,100 +1258,144 @@ eprintf ("++ EFL = 0x%08x  %d\n", ctx.EFlags, r_offsetof (CONTEXT, EFlags));
...
+			float *f=(float *)&fpregs.st_space;
+			c=c+(i*4);
+			f=f+(i*4);
+			eprintf ("st%d =%0.3lg (0x%016"PFMT64x") | %0.3f (0x%08x)  | %0.3f (0x%08x)\n"
+				,i ,d[0] ,b[0] ,f[0] ,c[0] ,f[1] ,c[1]);
```
Comment (ret2libc) [inline libr/debug/p/debug_native.c]: space after comma

**Comment by ret2libc** [comment]
I'm not sure about the #if/#else/#endif indentation. Usually they are not indented and they also make the code one level deeper. What do others think about indenting preprocessor directives?

**Comment by alvarofe** [comment]
I  did the indentation to make it more readable. i think is not the standard although the compiler doesn't complain. I prefer the indentation.

**Comment by radare** [comment]
Please squash for updated review.

**Comment by radare** [comment]
We shouldnt need to indent cpp ifs because all this stuff must be splitted in different files or reorganize it in a samer way. Thats a bit of the refactor i started a while ago under native/

**Comment by radare** [comment]
But correct cpp indwntation is like # always at first column and then the keyword after N spaces/tabs

**Comment by radare** [comment]
Some compilers dislike cpp macros not starting at first column

**Comment by alvarofe** [comment]
Yeah , definitely all this must be splitted in different files because now is a mess. As soon as I get a little time I will change it, the coding-styles, the macros ... and i will update the PR to comment more

**Comment by XVilka** [comment]
@alvarofe please, also extend this part, regarding the adding new architecture/system for the native debug - https://github.com/radare/radare2/wiki/Implementing-a-new-architecture#write-a-debugger-plugin

**Comment by alvarofe** [comment]
@XVilka of course :)

**Comment by radare** [comment]
merge it, lgtm and thats the best way to detect regressions :) works fine on linux and osx, need to test on android and such, bsd testers are welcome

**Comment by radare** [comment]
ok, didnt saw the comments. those comments needs to be addressed

**Comment by alvarofe** [comment]
I'm still working but I think that xnu would be finished except for some refinements maybe. @radare I will need you tell me what code could be erased like: #if 0 #endif, comments, old functions ... to let it in a clean state. 

The following days I will split the linux,bsd.

I have not applied any code-style yet so not be bad with me xD

**Comment by radare** [comment]
Try to use the standard code style from r2. I'll try to hack on clangformat to get it working as we want

**Comment by radare** [comment]
i merged another PR that may be problematic for this pr to be merged. Please, dont let pullreqs much time in the open state :) i prefer small/progressive/quickly to review and test changes than huge refactorings. Single squashed commits also help on this sometimes.

**Comment by radare** [comment]
I'm merging this, its not problematic, but is better to do a progressive refactoring than keep a pullreq open for a week. You can continue the refactoring in a new pullreq on top of the current master

**Comment by radare** [comment]
Thanks!

**Pull request closed without merging**